client/compliance_client.go

```diff
@@ -585,6 +654,894 @@ func (c *complianceGRPCClient) CancelOperation(ctx context.Context, req *longrun
+	if err != nil {
+		return nil, err
+	}
+	// Set the headers
+	for k, v := range c.xGoogMetadata {
+		httpReq.Header[k] = v
+	}
+	httpReq.Header["Content-Type"] = []string{"application/json"}
+
+	httpRsp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return nil, err
+	}
+	defer httpRsp.Body.Close()
+
+	if httpRsp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf(httpRsp.Status)
+	}
+
+	buf, err := ioutil.ReadAll(httpRsp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	unm := protojson.UnmarshalOptions{AllowPartial: true, DiscardUnknown: true}
+	rsp := &genprotopb.RepeatResponse{}
+
+	return rsp, unm.Unmarshal(buf, rsp)
+}
+
+// RepeatDataBodyInfo this method echoes the ComplianceData request. This method exercises
+// sending the a message-type field in the REST body. Per AIP-127, only
+// top-level, non-repeated fields can be sent this way.
+func (c *complianceRESTClient) RepeatDataBodyInfo(ctx context.Context, req *genprotopb.RepeatRequest, opts ...gax.CallOption) (*genprotopb.RepeatResponse, error) {
+	m := protojson.MarshalOptions{AllowPartial: true, EmitUnpopulated: true}
+	body := req.GetInfo()
+	jsonReq, err := m.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+
+	baseUrl, _ := url.Parse(c.endpoint)
+	baseUrl.Path += fmt.Sprintf("/v1beta1/repeat:bodyinfo")
+
+	params := url.Values{}
+	if req.GetName() != "" {
+		params.Add("name", fmt.Sprintf("%v", req.GetName()))
+	}
+	if req.GetServerVerify() {
+		params.Add("serverVerify", fmt.Sprintf("%v", req.GetServerVerify()))
+	}
+
+	baseUrl.RawQuery = params.Encode()
+
+	httpReq, err := http.NewRequestWithContext(ctx, "POST", baseUrl.String(), bytes.NewReader(jsonReq))
+	if err != nil {
+		return nil, err
+	}
+	// Set the headers
+	for k, v := range c.xGoogMetadata {
+		httpReq.Header[k] = v
+	}
+	httpReq.Header["Content-Type"] = []string{"application/json"}
+
+	httpRsp, err := c.httpClient.Do(httpReq)
+	if err != nil {
+		return nil, err
+	}
+	defer httpRsp.Body.Close()
+
+	if httpRsp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf(httpRsp.Status)
+	}
+
+	buf, err := ioutil.ReadAll(httpRsp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	unm := protojson.UnmarshalOptions{AllowPartial: true, DiscardUnknown: true}
+	rsp := &genprotopb.RepeatResponse{}
+
+	return rsp, unm.Unmarshal(buf, rsp)
+}
+
+// RepeatDataQuery this method echoes the ComplianceData request. This method exercises
+// sending all request fields as query parameters.
+func (c *complianceRESTClient) RepeatDataQuery(ctx context.Context, req *genprotopb.RepeatRequest, opts ...gax.CallOption) (*genprotopb.RepeatResponse, error) {
+	m := protojson.MarshalOptions{AllowPartial: true, EmitUnpopulated: true}
+	jsonReq, err := m.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	baseUrl, _ := url.Parse(c.endpoint)
+	baseUrl.Path += fmt.Sprintf("/v1beta1/repeat:query")
+
+	params := url.Values{}
+	if req.GetInfo().GetFBool() {
```

Comment:
I'm OK with this structure for expediency, but I think long term it might be desirable to use proto3 reflection to construct the list of query params. That will make the code more readable, and will allow us to handle recursive message definitions (which is both elegant and, I believe, allowed)

Reply:
The Showcase server is doing the opposite (json to proto) in a couple of places, like [this](https://github.com/googleapis/gapic-showcase/blob/04cab450b3ad6b7519ec7754db40fe633d7f47a6/util/genrest/resttools/populatefield.go#L63). That may be helpful for you. In particular, the [protoreflect](https://pkg.go.dev/google.golang.org/protobuf/reflect/protoreflect) package.

Reply:
That's extremely helpful. Thanks!